### PR TITLE
fix(web): optimized rollup output logic

### LINKED ISSUE #8475: @nrwl/web:rollup produces the wrong output
## Current Behavior
When using the following configuration:

```json
    "build": {
      "executor": "@nrwl/web:rollup",
      "outputs": [
        "{options.outputPath}"
      ],
      "options": {
        "outputPath": "dist/packages/falso",
        "tsConfig": "packages/falso/tsconfig.lib.json",
        "project": "packages/falso/package.json",
        "entryFile": "packages/falso/src/index.ts",
        "format": [
          "esm",
          "cjs"
        ]
      }
    },
```

The build output in `package.json` is:

```json
  "main": "./index.umd.js",
  "module": "./index.esm.js",
  "typings": "./index.d.ts"
```

## Expected Behavior

```json
  "main": "./index.cjs.js",
  "module": "./index.esm.js",
  "typings": "./index.d.ts"
```

### LINKED ISSUE #8505: Generated package.json should honour main and module fields
## Description
For now, it seems like the `main` and `module` fields of the generated `package.json` file are hardcoded to be the following format:

* `[package-name].umd.js`, and
* `[package-name].esm.js`

There are no ways to customise this, and it may cause issues especially when using the `preserveModules: true` option. When the entry file is set to be `index.ts`, for example, the built entry file will still be `index.js` even though the main/module fields refer to a file named after the package name + module type.

The `package.json` references may look like this:

```
{
  "name": "my-component-library",
  "version": "0.0.1",
  "main": "dist/index.js",
  "module": "dist/index.js",
  "typings": "dist/index.d.ts",
}
```

Yet, the generated `package.json` looks like this:

```
{
  "name": "my-component-library",
  "version": "0.0.1",
  "main": "./my-component-library.umd.js",
  "module": "./my-component-library.esm.js",
  "typings": "./index.d.ts",
}
```

while the file directory in the `dist/` folder looks like:

```
dist/
├── index.js
└── index.d.ts
```

## Motivation
This generally doesn't cause an issue since the consuming application, when encountering a non-existent `main` or `module` file, will simply fallback to look for `index.js` as a fail-safe behavior. However, we should still be allowed to customise the values of these fields.

There are also several use-cases on StackOverflow that highlights this need is not specifically relevant to me, but also a few others:

* [nx - configure main/module for generated package.json](https://stackoverflow.com/questions/69905141/nx-configure-main-module-for-generated-package-json)
* [@nrwl/web:package - output CJS](https://stackoverflow.com/questions/65857488/nrwl-webpackage-output-cjs)

## Suggested Implementation
Reconsider how the `main` and `module` fields are being written. Right now it is hardcoded to the package name: see lines here https://github.com/nrwl/nx/blob/9d78629ec021b21179c38e52f545c233dca6f25c/packages/web/src/executors/package/package.impl.ts#L302-L303

## Alternate Implementations
Allow us to define the `main` and `module` fields at the very least.

### PR DESCRIPTION
**Pull request by xiejay97**
ISSUES CLOSED: #8475
ISSUES CLOSED: #8505

## Related Issue(s)

Fixes #

### CODE REVIEW

File: packages/web/src/executors/rollup/rollup.impl.ts [hunk 1]
```diff
@@ -329,9 +330,35 @@ function updatePackageJson(
     /\.[jt]sx?$/,
     '.d.ts'
   );
-  packageJson.main = entryFileTmpl.replace('<%= extension %>', 'umd');
-  packageJson.module = entryFileTmpl.replace('<%= extension %>', 'esm');
-  packageJson.typings = `./${typingsFile}`;
+
+  // Update main field
+  if (!('main' in packageJson)) {
```
Comment (jaysoo) [inline packages/web/src/executors/rollup/rollup.impl.ts:335]: Simple as `!packageJson.main`?

File: packages/web/src/executors/rollup/rollup.impl.ts [hunk 2]
```diff
@@ -329,9 +330,35 @@ function updatePackageJson(
...
+    let extension: 'umd' | 'cjs' | null = null;
+    for (const format of options.format) {
```
Comment (jaysoo) [inline packages/web/src/executors/rollup/rollup.impl.ts:337]: This is simpler, and prefers `cjs` over `umd` which is more suitable.

```

  if (options.format.includes('cjs')) {
    packageJson.main = entryFileTmpl.replace('<%= extension %>', 'cjs');
  } else if (options.format.includes('umd')) {
    packageJson.main = entryFileTmpl.replace('<%= extension %>', 'umd');
  }
```

**Review by jaysoo — CHANGES_REQUESTED** [review]

**Comment by AgentEnder** [comment]
@xiejay97 Do you care to run `nx format` and re-push your branch? If we can get CI green on this, it would be good to get merged.

**Comment by xiejay97** [comment]
> @xiejay97 Do you care to run `nx format` and re-push your branch? If we can get CI green on this, it would be good to get merged.

OK now

**Comment by AndyClausen** [comment]
One thing to note here:
If I have a package A with tests that depend on package B, I need to set the `main` field in package B to `index.ts` or the tests in pakcage A will fail.
After this PR is merged, if I set the `main` field, it won't be overwritten when running rollup and I will be "unknowingly" publishing with the incorrect `main` field.
Maybe this is how it should work though, and the problem lies with Jest. I thought it was worth mentioning nonetheless.

**Comment by jaysoo** [comment]
This is fixed in latest Nx version (14.1.9)

**Pull request closed without merging**